### DISCUSSION TITLE
Tilt correction for huion

### DISCUSSION THREAD

**Discussion by retrotails** [kurikaesu/userspace-tablet-driver-daemon#79] [General]
I have a Kamvas Pro 16 (which otherwise works great with this driver, including the buttons and touch slider), likely near-identical to the 13" version. Firmware shows up as "HUION_M183_190521".
I noticed two issues related to the cursor position when tilting the pen, which led me to find this driver, and it has the same problems as the other options I've tried.

Issue 1 is, the cursor strays from the pen tip when it's tilted toward ordinal directions, while cardinal directions are fine.
I believe this is because when the pen is pointed at a corner, the tilt axes are both maxed out, which leads to a square shape, and the tablet firmware doesn't properly correct this into a circle.

https://github.com/kurikaesu/userspace-tablet-driver-daemon/assets/10767978/6a56de22-4852-410a-a24f-097c0c05ef87

Issue 2 is weird, it seems like the tablet just completely stops updating the tilt correction as soon as it starts a stroke, and doesn't resume until you lift it.
This means when you start a line, then change the pen tilt during the stroke, the cursor goes completely away from the pen tip. As soon as you lift the pen, the cursor snaps back to the correct position.

https://github.com/kurikaesu/userspace-tablet-driver-daemon/assets/10767978/eda1f48d-8a70-4fb9-874e-413b094e57b5

I made some simple tweaks that seem to fix it 90% for my tablet; the only issue I've noticed is the very tail of a stroke can snap to the wrong position making a spike, might be an easy fix.

```diff
diff --git a/src/huion_tablet.cpp b/src/huion_tablet.cpp
index f6b2d11..00aea31 100644
--- a/src/huion_tablet.cpp
+++ b/src/huion_tablet.cpp
@@ -389,11 +389,21 @@ void huion_tablet::handleDigitizerEventV1(libusb_device_handle *handle, unsigned
     uinput_send(uinputPens[handle], EV_SYN, SYN_REPORT, 1);
 }
 
+int offset_x = 0;
+int offset_y = 0;
+
+int old_tiltx = 0;
+int old_tilty = 0;
+
 void huion_tablet::handleDigitizerEventV2(libusb_device_handle *handle, unsigned char *data, size_t dataLen) {
     // Extract the X and Y position
     int penX = (data[8] << 16) + (data[3] << 8) + data[2];
     int penY = (data[5] << 8) + data[4];
 
+    // Grab the tilt values
+    short tiltx = (char)data[10];
+    short tilty = (char)data[11];
+
     // Check to see if the pen is touching
     std::bitset<8> stylusTipAndButton(data[1]);
     int pressure = (data[7] << 8) + data[6];
@@ -401,13 +411,25 @@ void huion_tablet::handleDigitizerEventV2(libusb_device_handle *handle, unsigned
     if (stylusTipAndButton.test(0)) {
         handlePenEnteredProximity(handle);
         handlePenTouchingDigitizer(handle, pressure);
+        
+        // Tablet stops correcting for tilt when pen touching begins, so we correct for tilt
+        // Notably, we do it the incorrect way the same as the tablet, because we correct it later
+        penX -=  (tiltx - old_tiltx) * 11;
+        penY -= -(tilty - old_tilty) * 11;
     } else {
         handlePenTouchingDigitizer(handle, pressure);
+        
+	old_tiltx = tiltx;
+	old_tilty = tilty;
     }
+    
+    // Tablet incorrectly corrects for tilt when hovering, this fixes the ordinal directions going too far
+    float s = fabs( sin( atan2(tilty,tiltx) * 2.0 ) );
+    float len = sqrt(tiltx*tiltx+tilty*tilty) / 60.0;
+    penX +=  tiltx * 5 * s * len;
+    penY += -tilty * 5 * s * len;
+    
 
-    // Grab the tilt values
-    short tiltx = (char)data[10];
-    short tilty = (char)data[11];
 
     // Check to see if the stylus buttons are being pressed
     if (stylusTipAndButton.test(1)) {

```

I'm posting here because I don't know how widespread this problem is. Maybe  it only affects some tablets, maybe there's a firmware update that fixes it, I don't know. Reviews online seem to indicate the Windows drivers don't have the issue. If anyone else has a Huion screen tablet, it'd be helpful to know if this is an issue.

**Comment by kurikaesu** [comment]
Wow! Super interesting. The first issue looks like the tilt values may be inverted.

The second one is plain weird. I'd be interested in seeing what the raw USB messages being produced looks like.

There is some commented out code in the huion driver that dumps the messages in the terminal if you were also curious to know.

**Comment by retrotails** [comment]
I think the tablet does tilt correction on it's own in firmware, before sending it over USB, so my code is more of a workaround for a hardware bug.

https://github.com/kurikaesu/userspace-tablet-driver-daemon/assets/10767978/bf2e8376-ad70-487b-a2bf-926b099a5d20

[huion_usb.txt](https://github.com/kurikaesu/userspace-tablet-driver-daemon/files/13432533/huion_usb.txt)
